openmicroscopy/openmicroscopy
# Api experimenters groups

# What this PR does

Adds initial support for ```/experimenters/``` and ```/experimentergroups/``` to the JSON API.

This includes updating the ```getQueryString()``` for the BlitzGateway ExperimenterWrapper and ExperimenterGroupWrapper classes, so that by default they load experimenters without groups and load groups without experimenters.

Need to use the opts to load experimenters / groups (same as for Image 'load_pixels' etc).
```
conn.getObjects("ExperimenterGroup", opts={load_experimenters=True})
conn.getObjects("Experimenter", opts={load_groups=True})
```
In several places in webadmin we are using ```conn.getObjects("ExperimenterGroup")``` to list groups (e.g. in the experimenter/new/ form) when we don't need to ```load_experimenters```. Now that these are not being loaded, the loading time for that page is ~2 x faster for me.

I've also updated the ```group.copyGroupExperimenterMap()``` and ```experimenter.copyGroupExperimenterMap()``` to lazy load if not previously loaded.
This means that doing the following is slower because of lazy loading, but at least it won't fail
```
for e in conn.getObjects("Experimenter"):
    e.copyGroupExperimenterMap()
    # or this also needs to load groups
    e.isAdmin()
```

TODOs:

 - [x] Support for single Experimenter or Group
 - [ ] Loading experimenterGroupMap etc?? (see below)
 - [ ] Query experimenters or groups by name? E.g. to look up your own experimenter ID from omeName or lastName? E.g. ```/experimenters/?omeName=will```

# Testing this PR

1. Browse to ```/api/v0/m/experimenters/``` to list all.
1. filter with ```?group=3``` or with  ```/api/v0/m/groups/3/experimenters```
1. Browse to ```/api/v0/m/experimentergroups/``` to list all.
1. filter groups with ```?experimenter=3``` or with  ```/api/v0/m/experimenters/3/experimentergroups```
1. Follow links to see single experimenter or group: ```/api/v0/m/experimenters/3/``` or ```/api/v0/m/groups/3/```
1. Need to test webclient and webadmin in places where we are loading experimenters / groups to check that we don't get unloaded exceptions:
    - Main /webclient/ page - Groups/Users menu
    - Search page (groups select option)
    - webadmin listing and editing groups
    - webadmin listing and editing experimenters
    - User settings in webclient (including for group leaders)


# Related reading

To discuss:
 - Do we expose the full name ```experimenterGroups``` anywhere (urls, filters etc) or do we simply stick with ```groups```?
 - Currently omero-marshal doesn't support loading groupExperimenterMap for Experimenters OR Groups. This means you can't choose to list Groups with Experimenters loaded (or list Experimenters with Groups loaded). Also, can't see which members are Owners of a Group.

cc @chris-allan @jburel
